visualization/app/codeCharta/ui/dialog/dialog.download.component.spec.ts

```diff
@@ -0,0 +1,298 @@
+import "./dialog.module"
+import "../codeMap/codeMap.module"
+import { DialogDownloadController, DownloadCheckboxNames } from "./dialog.download.component"
+import { SettingsService } from "../../state/settingsService/settings.service"
+import { FileStateService } from "../../state/fileState.service"
+import { CodeMapPreRenderService } from "../codeMap/codeMap.preRender.service"
+import { Settings, AttributeTypes, AttributeTypeValue } from "../../codeCharta.model"
+import { instantiateModule, getService } from "../../../../mocks/ng.mockhelper"
+import { stubDate } from "../../../../mocks/dateMock.helper"
+import {
+	DEFAULT_SETTINGS,
+	FILE_STATES,
+	VALID_NODE_WITH_PATH_AND_EXTENSION,
+	FILE_META,
+	VALID_EDGES,
+	BLACKLIST,
+	MARKED_PACKAGES
+} from "../../util/dataMocks"
+import _ from "lodash"
+
+describe("DialogDownloadController", () => {
+	stubDate(new Date(Date.UTC(2018, 11, 14, 9, 39)))
+	const newDate: string = "2018-12-14_09-39"
+
+	let dialogDownloadController: DialogDownloadController
+	let $mdDialog
+	let codeMapPreRenderService: CodeMapPreRenderService
+	let settingsService: SettingsService
+	let fileStateService: FileStateService
+	let settings: Settings
+
+	beforeEach(() => {
+		restartSystem()
+		withMockedSettingsService()
+		withMockedFileStateService()
+		withMockedCodeMapPreRenderService()
+		rebuildController()
+	})
+
+	function restartSystem() {
+		instantiateModule("app.codeCharta.ui.dialog")
+
+		$mdDialog = getService("$mdDialog")
+		codeMapPreRenderService = getService<CodeMapPreRenderService>("codeMapPreRenderService")
+		settingsService = getService<SettingsService>("settingsService")
+		fileStateService = getService<FileStateService>("fileStateService")
+
+		settings = _.cloneDeep(DEFAULT_SETTINGS)
+	}
+
+	function rebuildController() {
+		dialogDownloadController = new DialogDownloadController($mdDialog, codeMapPreRenderService, settingsService, fileStateService)
+	}
+
+	function withMockedCodeMapPreRenderService() {
+		codeMapPreRenderService = jest.fn<CodeMapPreRenderService>(() => {
+			return {
+				getRenderMap: jest.fn().mockReturnValue(VALID_NODE_WITH_PATH_AND_EXTENSION),
+				getRenderFileMeta: jest.fn().mockReturnValue(FILE_META)
+			}
+		})()
+	}
+
+	function withMockedSettingsService(mockSettings: Settings = settings) {
+		settingsService = jest.fn<SettingsService>(() => {
+			return {
+				getSettings: jest.fn().mockReturnValue(mockSettings)
+			}
+		})()
+	}
+
+	function withMockedFileStateService() {
+		fileStateService = jest.fn<FileStateService>(() => {
+			return {
+				getFileStates: jest.fn().mockReturnValue(FILE_STATES)
+			}
+		})()
+	}
+
+	function getFilteredFileContent(name: DownloadCheckboxNames) {
+		return dialogDownloadController["_viewModel"].fileContent.find(x => x.name == name)
+	}
+
```

Comment:
Can we get a wrapper here that all these tests happen on construction? You can spot that by seeing that only `rebuildController` is called as an act, but it would be great to know this by reading it from top to bottom.

```
describe("constructor", () => {
    describe("amoutOfAttributeTypes", () => {
        it("should be set correctly", () => {}
    }
}
```

etc

Reply:
Do you think of one describe-block with the string "constructor" that wraps all tests?

Reply:
Yes, exactly. And rename the other describe blocks to the field of the viewmodel you're currently testing.

Reply:
Did it.